tree/ntuple/v7/inc/ROOT/RColumn.hxx

```diff
@@ -81,26 +89,52 @@ public:
    void Connect(DescriptorId_t fieldId, RPageStorage *pageStorage);
 
    void Append(const RColumnElementBase &element) {
-      void *dst = fHeadPage.TryGrow(1);
-      if (dst == nullptr) {
-         Flush();
-         dst = fHeadPage.TryGrow(1);
-         R__ASSERT(dst != nullptr);
+      void *dst = fHeadPage[fHeadPageIdx].TryGrow(1);
+      R__ASSERT(dst != nullptr);
+      if (fHeadPage[fHeadPageIdx].GetNElements() == fApproxNElementsPerPage / 2) {
+         // Current page is at 50% fill level, we can now commit the previously used page
+         auto otherIdx = (fHeadPageIdx + 1) % 2;
+         if (fHeadPage[otherIdx].GetNElements())
+            fPageSink->CommitPage(fHandleSink, fHeadPage[otherIdx]);
       }
+
       element.WriteTo(dst, 1);
       fNElements++;
+
+      if (fHeadPage[fHeadPageIdx].GetNElements() == fApproxNElementsPerPage) {
+         // We are at 100% fill level, switch to the other head page
+         fHeadPageIdx = (fHeadPageIdx + 1) % 2;
+         fHeadPage[fHeadPageIdx].Reset(fNElements);
+      }
    }
 
    void AppendV(const RColumnElementBase &elemArray, std::size_t count) {
-      void *dst = fHeadPage.TryGrow(count);
-      if (dst == nullptr) {
+      if (fHeadPage[fHeadPageIdx].GetNElements() + count > fApproxNElementsPerPage) {
```

Comment:
The single-element `Append()` already takes care of page swapping when necessary.  As a possible optimization, one could use fewer calls of `AppendV()` in the loop instead of `count` calls to `Append()`.